tests/lms/services/_helpers/canvas_api_test.py

```diff
@@ -48,17 +55,118 @@ def test_public_url_request(self, ai_getter, route_url):
+
+    def test_it_raises_CanvasAPIError_if_the_response_is_invalid(
+        self, helper, prepared_request, raise_from, Schema
+    ):
+        Schema.return_value.parse.side_effect = ValidationError("error message")
+
+        with pytest.raises(CanvasAPIError, match="test_error_message") as exc_info:
+            helper.validated_response(prepared_request, Schema)
+
+        raise_from.assert_called_once_with(Schema.return_value.parse.side_effect)
+
+    def test_it_skips_validation_if_no_schema_is_given(
+        self, helper, prepared_request, Schema
+    ):
+        helper.validated_response(prepared_request)
+
+        Schema.assert_not_called()
+
     @pytest.fixture
-    def ai_getter(self, ai_getter):
-        ai_getter.developer_key.return_value = "test_developer_key"
-        ai_getter.developer_secret.return_value = "test_developer_secret"
-        ai_getter.lms_url.return_value = "https://my-canvas-instance.com/"
-        return ai_getter
+    def prepared_request(self):
+        return mock.sentinel.prepared_request
+
+    @pytest.fixture(autouse=True)
+    def requests(self, patch):
+        return patch("lms.services._helpers.canvas_api.requests")
 
     @pytest.fixture
-    def route_url(self, pyramid_request):
-        return pyramid_request.route_url
+    def requests_session(self, requests):
+        """The requests.Session object."""
+        return requests.Session.return_value
 
     @pytest.fixture(autouse=True)
-    def routes(self, pyramid_config):
-        pyramid_config.add_route("canvas_oauth_callback", "/canvas_oauth_callback")
+    def requests_response(self, requests_session):
+        """The requests.Response object returned by requests's send() method."""
+        requests_response = requests_session.send.return_value
+        requests_response.status_code = 200
+        requests_response.reason = "OK"
+        requests_response.text = ""
+        return requests_response
+
+    @pytest.fixture
+    def Schema(self):
+        return mock.create_autospec(PyramidRequestSchema, spec_set=True)
+
+
+@pytest.fixture
+def ai_getter(ai_getter):
+    ai_getter.developer_key.return_value = "test_developer_key"
+    ai_getter.developer_secret.return_value = "test_developer_secret"
+    ai_getter.lms_url.return_value = "https://my-canvas-instance.com/"
+    return ai_getter
+
+
+@pytest.fixture
+def helper(ai_getter, route_url):
+    return CanvasAPIHelper("test_consumer_key", ai_getter, route_url)
+
+
+@pytest.fixture(autouse=True)
+def raise_from(request):
+    # Always replace the original CanvasAPIError.raise_from() after each test.
+    original = CanvasAPIError.raise_from
+
+    def finalizer():
+        CanvasAPIError.raise_from = original
+
+    request.addfinalizer(finalizer)
+
+    # Replace CanvasAPIError.raise_from() with a mocl.
+    # This is done manually because using the normal mock.patch() or even
+    # mock.patch.object() to patch this static method didn't seem to work.
```

Comment:
Thanks for the clarification. Out of interest what was the error/issue?

Reply:
It appeared that mock replaced the static method with a non-callable mock. I didn't investigate for long, to be honest